File: solutions/0515-find-largest-value-in-each-tree-row/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function(root) {
    const output = []
    const queue = [root];
    
    while(queue.length > 0){
        const s = queue.length;
        let max;
        for(let i = 0; i < s; i++){
            const node = queue.shift();
            if(node === null){
                continue;
            }
            
            if(max === undefined || node.val > max){
                max = node.val
            }
            if(node.left){
                queue.push(node.left);
            }
            if(node.right){
                queue.push(node.right);
            }
            
        }
        
        if(max !== undefined){
            output.push(max)
        }
    }
    
    return output;
};

File: solutions/0515-find-largest-value-in-each-tree-row/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        queue = deque()
        queue.append(root)
        output = []
        
        while queue:
            max = None
            l = len(queue)
            for _ in range(l):
                node = queue.popleft()
                if node is None:
                    continue
                
                if max is None or node.val > max:
                    max = node.val
                    
                if node.left:
                    queue.append(node.left)
                    
                if node.right:
                    queue.append(node.right)
                    
            if max is not None:
                output.append(max)
        
        return output
